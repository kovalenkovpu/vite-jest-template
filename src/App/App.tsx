import * as React from "react";

import logo from "../assets/logo.svg";
import "./App.css";

import { useHandleCount } from "../hooks/useHandleCount";

function App() {
  const { count, increaseCount } = useHandleCount(0);

  return (
    <div className="App">
      <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
        <p>Hello Vite + React!</p>
        <p>
          <button type="button" onClick={increaseCount}>
            count is: {count}
          </button>
        </p>
        <p>
          Edit <code>App.tsx</code> and save to test HMR updates.
        </p>
      </header>
    </div>
  );
}

export default App;
