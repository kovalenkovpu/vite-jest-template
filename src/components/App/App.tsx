import * as React from "react";

import logo from "../../assets/logo.svg";
import classes from "./App.module.scss";

import { useHandleCount } from "../../hooks/useHandleCount";
import { Header } from "../Header/Header";

const App: React.VFC = () => {
  const { count, increaseCount } = useHandleCount(0);

  return (
    <div className={classes.app}>
      <Header />
      <header className={classes.appHeader}>
        <img src={logo} className={classes.appLogo} alt="logo" />

        <p>
          <button
            className={classes.button}
            type="button"
            onClick={increaseCount}
          >
            count is: {count}
          </button>
        </p>
        <p>
          Edit <code>App.tsx</code> and save to test HMR updates.
        </p>
      </header>
    </div>
  );
};

export default App;
