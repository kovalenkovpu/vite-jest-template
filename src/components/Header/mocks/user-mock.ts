import { rest } from "msw";

import { UserData } from "../../../services/user-service";

const userMockData: UserData = {
  username: "John Doe",
  email: "john_doe@gmail.com",
};

const userMocks = [
  rest.get<undefined, UserData>("/api/user", (req, res, ctx) => {
    return res(ctx.status(200), ctx.delay(1000), ctx.json(userMockData));
  }),
];

export { userMocks, userMockData };
