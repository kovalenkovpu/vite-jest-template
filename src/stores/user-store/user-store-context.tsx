import * as React from "react";

import { DIContainer } from "../../inversify.config";
import { TYPES } from "../../inversify.types";

import { IUserStore } from "./user-store";

const userStore = DIContainer.get<IUserStore>(TYPES.UserStore);

const UserStoreContext = React.createContext<IUserStore>(userStore);

const UserStoreContextProvider: React.FC = ({ children }) => (
  <UserStoreContext.Provider value={userStore}>
    {children}
  </UserStoreContext.Provider>
);

export { UserStoreContext, UserStoreContextProvider };
