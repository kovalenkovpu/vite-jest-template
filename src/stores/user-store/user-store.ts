import { injectable, inject } from "inversify";
import "reflect-metadata";
import { action, makeObservable, observable, runInAction } from "mobx";

import { UserData, UserService } from "../../services/user-service";
import { TYPES } from "../../inversify.types";

interface IUserStore {
  user: UserData | null;
  isDataLoading: boolean;

  fetchUserData: () => Promise<void>;
}

@injectable()
class UserStore implements IUserStore {
  user: UserData | null;
  isDataLoading: boolean;

  private userService: UserService;

  constructor(@inject(TYPES.UserService) userService: UserService) {
    this.user = null;
    this.isDataLoading = false;
    this.userService = userService;

    makeObservable(this, {
      user: observable,
      isDataLoading: observable,
      fetchUserData: action,
    });

    // Above code is just the example, can be simply:
    // makeAutoObservable(this);
  }

  fetchUserData = async () => {
    this.isDataLoading = true;

    try {
      const data = await this.userService.fetchUserData();

      // Otherwise we have warning about mutating the state outside of the action
      runInAction(() => {
        this.user = data;
      });
    } catch (error) {
    } finally {
      runInAction(() => {
        this.isDataLoading = false;
      });
    }
  };
}

export type { IUserStore };
export { UserStore };
