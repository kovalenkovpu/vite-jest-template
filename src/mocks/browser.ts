// src/mocks/browser.js
import { setupWorker } from "msw";
import { userMocks } from "../components/Header/mocks/user-mock";

// This configures a Service Worker with the given request handlers.
export const worker = setupWorker(...userMocks);
