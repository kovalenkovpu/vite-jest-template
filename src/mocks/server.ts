import { setupServer } from "msw/node";

import { userMocks } from "../components/Header/mocks/user-mock";

export const server = setupServer(...userMocks);
