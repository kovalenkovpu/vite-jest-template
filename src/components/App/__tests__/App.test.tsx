import * as React from "react";
import { rest } from "msw";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import App from "../App";
import { UserStoreContextProvider } from "../../../stores/user-store/user-store-context";
import { userMockData } from "../../Header/mocks/user-mock";
import { UserData } from "../../../stores/user-store/user-store";
import { server } from "../../../mocks/server";

describe("App", () => {
  test("should change the text content when clicked on the button", () => {
    render(<App />);

    expect(screen.getByRole("button")).toHaveTextContent("0");

    const button = screen.getByRole("button");

    fireEvent.click(button);
    fireEvent.click(button);

    expect(screen.getByRole("button")).toHaveTextContent("2");
  });

  test("should match the snapshot when user data is not loaded", () => {
    const { asFragment } = render(<App />);

    expect(asFragment()).toMatchSnapshot();
  });

  test("should contain proper user info when user data is loaded", async () => {
    server.use(
      rest.get<undefined, UserData>("/api/user", (req, res, ctx) => {
        return res(ctx.status(200), ctx.json(userMockData));
      })
    );

    const wrapper: React.FC = ({ children }) => (
      <UserStoreContextProvider>{children}</UserStoreContextProvider>
    );

    render(<App />, { wrapper });

    const updatedUserDataText = new RegExp(
      `Hello ${userMockData.username}`,
      "i"
    );

    waitFor(() =>
      expect(screen.getByText(updatedUserDataText)).toBeInTheDocument()
    );
  });
});
