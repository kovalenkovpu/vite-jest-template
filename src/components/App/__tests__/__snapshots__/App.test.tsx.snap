// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="app"
  >
    <section
      class="headerContainer"
    >
      <section>
        User data is loading...
      </section>
    </section>
    <header
      class="appHeader"
    >
      <img
        alt="logo"
        class="appLogo"
        src="test-file-stub"
      />
      <p>
        <button
          class="button"
          type="button"
        >
          count is: 0
        </button>
      </p>
      <p>
        Edit 
        <code>
          App.tsx
        </code>
         and save to test HMR updates.
      </p>
    </header>
  </div>
</DocumentFragment>
`;
